[DRAFT] INitial version of sticky export

tugsbayasgalan opened the pull request:
Summary: This is to make torchnative demos and benchmarking real models more simple by not requiring ppl to find example inputs first.

Test Plan: CI

Differential Revision: D72815584

tugsbayasgalan commented on torch/export/experimental/__init__.py:
```diff
@@ -67,3 +69,58 @@ def _export_forward_backward(
     return ep._update(gm, new_graph_signature)
+
+
+def _sticky_export(model):
```
I intentinally didn't make it a decorator because that would mean we need to change user code. Can extend it for decorator tho, if ppl find it more useful.

tugsbayasgalan commented on torch/export/experimental/__init__.py:
```diff
@@ -67,3 +69,58 @@ def _export_forward_backward(
+            model._exported_module = export(model, args, kwargs, dynamic_shapes=shapes_collection, strict=False).module()
+        return model._exported_module(*args, **kwargs)
+
+    model.forward = exported_forward
```
The idea is that export runtime assertion pass will fire when there is input mismatch

zhxchen17 replied:
@tugsbayasgalan I didn't see this get tested. Can you add unittest to ensure this doesn't have soundness issue over shapes?

tugsbayasgalan replied:
sure!

tugsbayasgalan commented on torch/export/experimental/__init__.py:
```diff
@@ -67,3 +69,58 @@ def _export_forward_backward(
+    return model
+
+
+def _undo_sticky_export(model):
```
Yeah! But this is probably needed if someone wants to undo sticky export

zhxchen17 replied:
I don't think it's a good idea to even expose an experimental interface to undo the decorator. we should rename the this to something like
```
def _undo_sticky_export_TESTING_ONLY
```

zhxchen17 reviewed (changes requested):
Tests and linter are failing

zhxchen17 commented on torch/export/experimental/__init__.py:
```diff
@@ -67,3 +71,33 @@ def _export_forward_backward(
+                    model,
+                    args,
+                    kwargs,
+                    strict=False
```
strict is already False by default?

zhxchen17 commented on torch/export/experimental/__init__.py:
```diff
@@ -67,3 +71,33 @@ def _export_forward_backward(
+                    args,
+                    kwargs,
+                    strict=False
+                ).module()
```
I wish we still could have dynamic shape... can you make this configurable?

zhxchen17 replied:
maybe let user pass something here and default can be None

zhxchen17 replied:
update: a dynamic_shape_callback should be fine.

tugsbayasgalan committed:
INitial version of sticky export (#151047)

Summary:

This is to make torchnative demos and benchmarking real models more simple by not requiring ppl to find example inputs first.

Test Plan: CI

Differential Revision: D72815584
